Rewrite ReactDOMComponentTree-test to test behavior using Public API

GordyD opened the pull request:
This is part of #11299

I've tried to identify cases where code within ReactDOMComponentTree is exercised via the public API and have updated accordingly but I'm not entirely sure whether I'm on the right track. I thought I'd PR to get feedback from the community. Looking forward to comments.

gaearon commented:
I think ideally instead of reading private fields and methods, we should try to test the *behavior* those fields and methods are used for. This would require some investigation, e.g. where is `getClosestInstanceFromNode` or `__reactEventHandlers` fields used. Then come up with a test that looks like application code that would trigger the corresponding behavior and verify its correctness.

Does this make sense?

GordyD commented:
@gaearon Makes sense. I'll investigate further this weekend. In the mean time, could you provide feedback on the test cases I added to give me an idea of whether I'm on the right track or not.

gaearon commented:
That was my feedback: while we could do it this way, it's still testing internals more than the actual behavior.

GordyD commented:
Would this be a plausible test for checking behavior that getInstanceFromNode is used for in renderSubtreeIntoContainer? (albeit only in DEV mode)

```
it('finds instance from node to stop rendering over other react rendered components', () => {
    spyOn(console, 'error');
    const component = <div><span>Hello</span></div>;
    const anotherComponent = <div></div>;
    const container = document.createElement('div');
    const instance = ReactDOM.render(component, container);
    ReactDOM.render(anotherComponent, instance);
    expectDev(console.error.calls.count()).toBe(1);
    expectDev(console.error.calls.argsFor(0)[0]).toContain(
      'render(...): Replacing React-rendered children with a new root ' +
      'component. If you intended to update the children of this node, ' +
      'you should instead have the existing children update their state ' +
      'and render the new components instead of calling ReactDOM.render.'
    );
  });
```

gaearon commented on packages/react-dom/src/__tests__/ReactDOMComponentTree-test.js:
```diff
@@ -68,52 +78,115 @@ describe('ReactDOMComponentTree', () => {
+      a = null;
+      _onChange = e => {
+        const node = e.currentTarget;
+        expect(node.value).toEqual(finishValue);
```
The problem with tests like this is that if `_onChange` never fires, the test passes (since none of the expectations were called). We need to move those out.

GordyD replied:
Thanks for the comment @gaearon . I'll update to move assertions out of the event handler - this is much simpler. 

Though FWIW from my understanding of [async tests in Jest](https://facebook.github.io/jest/docs/en/asynchronous.html#callbacks) if `done()` is never called, due to `_onChange` not firing, then it would fail - albeit the failure message would not be the most helpful.

GordyD commented:
@gaearon I've updated the diff, can I get another review?

gaearon commented on packages/react-dom/src/__tests__/ReactDOMComponentTree-test.js:
```diff
@@ -10,32 +10,42 @@
+    const inputEvent = new Event('input', {
+      bubbles: true,
+    });
+    setUntrackedInputValue.call(elem, value);
```
Can you remind me why this is necessary?

GordyD replied:
In order to simulate the behavior of trying to switch a component from being uncontrolled to controlled we need to assign a new value to the input without it being tracked - I found this [approach](https://github.com/facebook/react/pull/11309#pullrequestreview-70996771) after reading through the comments of #11309.

gaearon commented on packages/react-dom/src/__tests__/ReactDOMComponentTree-test.js:
```diff
@@ -10,32 +10,42 @@
-    // TODO: can we express this test with only public API?
-    ReactDOMComponentTree = require('../client/ReactDOMComponentTree');
     ReactDOMServer = require('react-dom/server');
+    document.innerHTML = '';
```
How do you feel about following this approach instead? Feels a tiny but more obvious what's going on and why IMO. https://github.com/facebook/react/pull/11517/commits/d7745d69917b28298d27b8e411329ca67cae4668

GordyD replied:
I like it. Tests are more focussed on testing and we move set up to where it should be - thanks for the pointer. I'll adopt this approach.

gaearon commented on packages/react-dom/src/__tests__/ReactDOMComponentTree-test.js:
```diff
@@ -68,56 +78,108 @@ describe('ReactDOMComponentTree', () => {
-    class Component extends React.Component {
+  it('finds closest instance for node when an event happens', () => {
+    const nonReactElemID = 'aID';
+    const innerHTML = {__html: `<div id="${nonReactElemID}"></div>`};
```
Smart.

GordyD commented:
I've updated the tests to adopt the set up and tear down approach demonstrated in d7745d6. 

After making changes I watched what happened as I commented out different branches of code in ReactDOMComponentTree to manually verify that the updated tests exercise and provide coverage for the module.

Notable Observations:

 - commenting out lines 26-28 of `getClosestInstanceFromNode` does not cause any failing tests, but that is expected due to the logic on line 32 - I validated this code path is exercised by the test though by logging here
 - commenting out line 47 in `getClosestInstanceFromNode` does not result in failing tests, but again because of the logic following this is to be expected - I validated this code path is exercised by the test by logging here
 - commenting out lines 49-51 in `getClosestInstanceFromNode` does not result in failing tests -tests do not currently exercise this code path
 - commenting out lines lines 66 or 69 do not result in any failing tests

All other blocks commented out resulted in failing tests.

I'm not too sure at this point how to add/update the tests to provide coverage for lines 49-51, 66 and 69 within ReactDOMComponentTree -  I'll continue to give it some thought this week.

gaearon commented:
It would be nice to try to cover these too. On the other hand, if the original tests didn't cover those either (can you check?) maybe it's okay to merge as is :-)

gaearon commented on packages/react-dom/src/__tests__/ReactDOMComponentTree-test.js:
```diff
@@ -68,56 +63,122 @@ describe('ReactDOMComponentTree', () => {
+    expect(currentTargetID).toBe(closestInstanceID);
+  });
+
+  it('finds a controlled instance from node and gets its current fiber props', () => {
```
If you want to test `getFiberCurrentPropsFromNode`, I think there are more assertions you might need to make. For example I can currently replace its implementation with

```js
return node[internalInstanceKey].memoizedProps;
```

and the test still passes. However, this is a wrong implementation (`node[internalInstanceKey]` can potentially point to an "alternate" fiber that is not "current". Current and alternate swap places on updates.)

Can we make this test more solid by tracing what happens with "current" props? e.g. maybe they're used in some way where alternate props would've been wrong. You can git blame when `getFiberCurrentPropsFromNode` was added to maybe find the cases that were broken before it.

GordyD replied:
Thanks for the detail, I'll check out the blame and track down ways to make the test more robust.

GordyD commented:
Checking out master and then commenting out those lines (49-51, 66, 69) do not result in failures - so it doesn't look like these lines are covered in the original test.

Another thing of note I just realized - commenting out body of `getNodeFromInstance` (line 80) does not result in the test that currently exists for this function! I'm going to look into re-writing the test for this so that we do exercise it. It does result in failures.

gaearon commented:
I appreciate the effort by the way. This is one of the harder ones. Hope you don't feel discouraged that it's not so fast.

GordyD commented:
Not at all, it's interesting to dig into - I'm learning a lot. 

P.S I might be a little slower to update this week as I'm on call, will update latest by the weekend.

gaearon commented:
I'll be on vacation this week so sounds good :-)

GordyD commented:
I've replaced the original test for `getNodeFromInstance` to exercise the function. I've also added a new test to test the implementation of `getFiberCurrentPropsFromNode` after following your pointer @gaearon  and checking out #8607. I found a test case that we could simplify to test behavior exhibited reliant upon its implementation. Swapping it out with `node[internalInstanceKey].memoizedProps` causes this case to fail.

gaearon commented:
You are awesome 👏
